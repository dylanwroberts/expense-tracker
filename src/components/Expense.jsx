import React from 'react';

export const Expense = props => (
    <div>
        <p>{props.expense.description}</p>
        <p>{props.expense.amount}</p>
    </div>
);

export default Expense;