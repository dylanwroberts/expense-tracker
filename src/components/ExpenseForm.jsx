import React from 'react';
import moment from 'moment';

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);

        //adds support for passing an expense in as the default
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            error: ''
        }
    }

    handleDescriptionChange = e => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    handleAmountChange = e => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }

    handleNoteChange = e => {
        const note = e.target.value;

    }

    handleFormSubmit = e => {
        e.preventDefault();

        if(!this.state.description) {
            this.setState(() => ({ error: 'You must provide a description.' }));
        }
        else if(!this.state.amount) {
            this.setState(() => ({ error: 'You must provide an amount.' }));
        }
        else {
            this.setState(() => ({ error: '' }));

            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()
            });
        }
    }

    render() {
        return (
            <div>
                <h1>Add a new expense</h1>
                { this.state.error ? <p>{this.state.error}</p> : null }
                <form>
                    <input type="text" placeholder="description" />
                    <input type="number" placeholder="amount" />
            
                    <textarea placeholder="note" />
                    <button>{this.props.buttonText}</button>
                </form>
            </div>
        );
    }
}