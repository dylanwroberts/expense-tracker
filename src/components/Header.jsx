import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <div>
        <h1>Expense Tracker</h1>
        <NavLink activeClassName="nav-link-active" to="/">Home</NavLink>
        <NavLink activeClassName="nav-link-active" to="/add">Add</NavLink>
    </div>
);

export default Header;