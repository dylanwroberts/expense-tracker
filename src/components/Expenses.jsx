import React from 'react';
import Expense from './Expense';
import { connect } from 'react-redux';
import getFilteredExpenses from '../selectors/getFilteredExpenses';

//named export
export const Expenses = props => (
    <>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map(expense => <Expense key={expense.description} expense={expense} />)
            )
        }
        <div>
        </div>
    </>
);

const mapStateToProps = state => ({
    expenses: getFilteredExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(Expenses);