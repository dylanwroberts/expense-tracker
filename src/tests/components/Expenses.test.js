import React from 'react';
import { shallow } from 'enzyme';
import { Expenses } from '../../components/Expenses';
import expenses from '../fixtures/expenses';

test('should render the Expenses component with expenses fixture', () => {
    const wrapper = shallow(<Expenses expenses={expenses} />);

    expect(wrapper).toMatchSnapshot();
});

test('should render Expenses component with no expenses', () => {
    const wrapper = shallow(<Expenses expenses={[]} />);
    
    expect(wrapper).toMatchSnapshot();
});