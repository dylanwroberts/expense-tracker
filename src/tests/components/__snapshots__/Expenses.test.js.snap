// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expenses component with no expenses 1`] = `
<Fragment>
  <p>
    No expenses
  </p>
  <div />
</Fragment>
`;

exports[`should render the Expenses component with expenses fixture 1`] = `
<Fragment>
  <Expense
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Food",
        "id": "1",
        "note": "",
      }
    }
    key="Food"
  />
  <Expense
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    key="Rent"
  />
  <Expense
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Eggs",
        "id": "3",
        "note": "",
      }
    }
    key="Eggs"
  />
  <div />
</Fragment>
`;
